iree/compiler/Codegen/Common/SetNumWorkgroupsFromLinalgExtPass.cpp

```diff
@@ -0,0 +1,129 @@
+        op->getResult(0).replaceAllUsesWith(idx);
+        op->erase();
+        return;
+      }
+      if (op->getName().getStringRef() == kFakeHALWorkgroupCountOpName) {
+        ImplicitLocOpBuilder b(op->getLoc(), op);
+        Value idx = b.create<IREE::HAL::InterfaceWorkgroupCountOp>(
+            op->getAttr("idx").cast<IntegerAttr>().getInt());
+        op->getResult(0).replaceAllUsesWith(idx);
+        op->erase();
+        return;
+      }
+      if (op->getName().getStringRef() == kFakeHALReturnOpName) {
+        ImplicitLocOpBuilder b(op->getLoc(), op);
+        b.create<IREE::HAL::ReturnOp>(op->getOperands());
+        op->erase();
+        return;
+      }
+    });
+
+    bool numWorkgroupIsSet = false;
+    assert(entryPointOp.workgroup_count_region().empty() &&
+           "Expected a single entryPoint op with no regions");
+
+    funcOp->walk([&](Operation *op) {
+      if (op->getName().getStringRef() == kFakeHALEntryPointRegionOpName) {
+        assert(op->getNumRegions() == 1 && "Expected single region op");
+        assert(!numWorkgroupIsSet);
+        numWorkgroupIsSet = true;
+        IRRewriter rewriter(op->getContext());
+        rewriter.setInsertionPoint(entryPointOp);
+        auto clonedEntryPointOp =
+            rewriter.create<IREE::HAL::ExecutableEntryPointOp>(
+                entryPointOp.getLoc(), entryPointOp.sym_nameAttr(),
+                entryPointOp.ordinalAttr(), entryPointOp.layoutAttr(),
+                entryPointOp.workgroup_sizeAttr(),
+                entryPointOp.workgroup_local_memoryAttr(), 1);
+        Block &block =
+            clonedEntryPointOp.workgroup_count_region().front().emplaceBlock();
+        rewriter.mergeBlocks(&op->getRegion(0).front(), &block);
+        // TODO: Don't add args post-hoc and instead replace them when
+        // kFakeHALEntryPointRegionOpName supports taking bbArgs itself.
+        block.addArgument(rewriter.getIndexType(), op->getLoc());
+        block.addArgument(rewriter.getIndexType(), op->getLoc());
+        block.addArgument(rewriter.getIndexType(), op->getLoc());
+        op->erase();
+        entryPointOp.erase();
+        clonedEntryPointOp->dump();
+      }
+    });
+  }
+
+  // Apply post distribution canonicalization passes.
+  RewritePatternSet canonicalization(context);
+  AffineMinOp::getCanonicalizationPatterns(canonicalization, context);
+  populateAffineMinSCFCanonicalizationPattern(canonicalization);
+  IREE::Flow::populateFlowDispatchCanonicalizationPatterns(canonicalization,
+                                                           context);
+  if (failed(
+          applyPatternsAndFoldGreedily(module, std::move(canonicalization)))) {
+    return signalPassFailure();
+  }
+}
+
+std::unique_ptr<OperationPass<IREE::HAL::ExecutableVariantOp>>
+createSetNumWorkgroupsFromLinalgExtPass() {
+  return std::make_unique<SetNumWorkgroupsFromLinalgExtPass>();
+}
+
+}  // namespace iree_compiler
+}  // namespace mlir
```

Comment:
Could we add some tests for this pass. The `hal.executable.entry_point` is scheduled to be cleaned up. So having lit tests will make sure I dont regress the functionality this pass is creating.

Reply:
tests were in the followup commit, I am folding the 2 together and will abandon this one.

Reply:
Just looking for a couple of simple lit tests that document what this pass does so that this path doesnt regress when things around `hal.executable.entry_point` changes.